### PR TITLE
[12.0][MIG]partner_contact_in_several_companies

### PR DESCRIPTION
**Pull request by kongrattapong**
done

### CODE REVIEW

File: partner_contact_in_several_companies/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```
Comment (NL66278) [inline partner_contact_in_several_companies/__init__.py]: Please use https: on all license lines.

File: partner_contact_in_several_companies/models/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
```
Comment (NL66278) [inline partner_contact_in_several_companies/models/__init__.py]: Coding not needed anymore (python3!)

File: partner_contact_in_several_companies/models/__init__.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from . import res_partner, ir_actions
```
Comment (NL66278) [inline partner_contact_in_several_companies/models/__init__.py:3]: Please each input on its own line.

File: partner_contact_in_several_companies/models/ir_actions.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
```
Comment (NL66278) [inline partner_contact_in_several_companies/models/ir_actions.py]: Also here, and everywhere, no more coding line.

File: partner_contact_in_several_companies/models/ir_actions.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+class IRActionsWindow(models.Model):
+    _inherit = 'ir.actions.act_window'
+
+    @api.multi
+    def read(self, fields=None, context=None, load='_classic_read'):
+        actions = super(IRActionsWindow, self).read(fields=fields, load=load)
+        for action in actions:
+            if action.get('res_model', '') == 'res.partner':
+                # By default, only show standalone contact
+                action_context = action.get('context', '{}') or '{}'
```
Comment (NL66278) [inline partner_contact_in_several_companies/models/ir_actions.py]: The `or '{}'` seems superfluous, as you already put that as default in the get.

File: partner_contact_in_several_companies/models/ir_actions.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+                if 'search_show_all_positions' not in action_context:
+                    action['context'] = action_context.replace(
+                        '{',
+                        ("{'search_show_all_positions': "
+                         "{'is_set': True, 'set_value': False},"),
+                        1)
+        return actions
```
Comment (NL66278) [inline partner_contact_in_several_companies/models/ir_actions.py]: It seems to me the preceding lines can not work, or am I missing something? Tried in ipython:
```
(buildoutenv3) openeyedev@laptop17 ~ $ ipython3
Python 3.5.2 (default, Nov 12 2018, 13:43:14) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.3.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: action_context = {}                                                                                                                                                  

In [2]: action_context.replace('{', ("{'search_show_all_positions': {'is_set': True, 'set_value': False},"), 1)                                                              
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-7d214fbc2633> in <module>
----> 1 action_context.replace('{', ("{'search_show_all_positions': {'is_set': True, 'set_value': False},"), 1)

AttributeError: 'dict' object has no attribute 'replace'
```

File: partner_contact_in_several_companies/models/res_partner.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+from odoo import api, fields, models, _
+from odoo.osv import expression
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    contact_type = fields.Selection(
+        [('standalone', _('Standalone Contact')),
```
Comment (NL66278) [inline partner_contact_in_several_companies/models/res_partner.py]: No need for _("something") in selection. Strings are automatically available for translation.

File: partner_contact_in_several_companies/models/res_partner.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+         ('attached', _('Attached to existing Contact')),
+         ],
+        compute='_compute_contact_type',
+        store=True,
+        index=True,
+        default='standalone')
+    contact_id = fields.Many2one('res.partner', string='Main Contact',
```
Comment (NL66278) [inline partner_contact_in_several_companies/models/res_partner.py]: Please use one style for indentation. Best always indent with four spaces, it makes for much nicer code then having everything crammed at the end of lines.

File: partner_contact_in_several_companies/models/res_partner.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+            attached_contact_args = expression.AND(
+                (args, [('contact_type', '=', 'attached')])
+            )
+            attached_contacts = super(ResPartner, self).search(
+                attached_contact_args)
+            args = expression.OR((
+                expression.AND(([('contact_type', '=', 'standalone')], args)),
+                [('other_contact_ids', 'in', attached_contacts.ids)],
+            ))
+        return super(ResPartner, self).search(args, offset=offset,
```
Comment (NL66278) [inline partner_contact_in_several_companies/models/res_partner.py:69]: Same here, always indent with four spaces.

**Review by NL66278 — CHANGES_REQUESTED** [review]
Thanks for your contribution!
Please take up the changes suggested in the comments.

File: partner_contact_in_several_companies/__init__.py (hunk 1 above)
Comment (kongrattapong) [reply partner_contact_in_several_companies/__init__.py]: Sorry for late replay I'll do it.

**Comment by kongrattapong** [comment]
> Thanks for your contribution!
> Please take up the changes suggested in the comments.

**Comment by rafaelbn** [comment]
Hi @kongrattapong , thanks for this contribution. Please

- This branch has conflicts that must be resolved
- @NL66278 make you a review, could you please attened comments?

Thanks

**Review by NL66278 — APPROVED** [review]
:+1:  Thanks for taking up suggestions. One thing: it might be that your branch needs a rebase, I see now a lot of changes to other modules that should not be there...

**Comment by kongrattapong** [comment]
>  Thanks for taking up suggestions. One thing: it might be that your branch needs a rebase, I see now a lot of changes to other modules that should not be there...

sorry I'm a new contributor, I try to do your suggest but I think I do something wrong, can you guide me what I need to do?

**Comment by kongrattapong** [comment]
> > Thanks for taking up suggestions. One thing: it might be that your branch needs a rebase, I see now a lot of changes to other modules that should not be there...
> 
> @NL66278  sorry I'm a new contributor, I try to do your suggest but I think I do something wrong, can you guide me what I need to do?

**Comment by NL66278** [comment]
@kongrattapong:
```
$ git checkout 12.0
$ git pull
$ git checkout 12.0-mig-partner_contact_in_several_companies
$ git rebase 12.0

Then resolve conflicts if any...
$ git push 12.0-mig-partner_contact_in_several_companies
```

**Comment by kongrattapong** [comment]
> @kongrattapong:
> 
> ```
> $ git checkout 12.0
> $ git pull
> $ git checkout 12.0-mig-partner_contact_in_several_companies
> $ git rebase 12.0
> 
> Then resolve conflicts if any...
> $ git push 12.0-mig-partner_contact_in_several_companies
> ```

**Comment by kongrattapong** [comment]
> @NL66278 
> 
> ```
> $ git checkout 12.0
> $ git pull
> $ git checkout 12.0-mig-partner_contact_in_several_companies
> $ git rebase 12.0
> 
> Then resolve conflicts if any...
> $ git push 12.0-mig-partner_contact_in_several_companies
> ```

I did as you suggested.

**Comment by kongrattapong** [comment]
@NL66278 I have some question. I want this PR is merge how can I do?

**Comment by NL66278** [comment]
@kongrattapong We have to wait some time for other reviews. Would help if you could encourage other people to do a review. When approved by runbot, I can merge.

**Comment by kongrattapong** [comment]
cc. @Saran440

**Comment by luffah** [comment]
Note : name attribute is missing for page 'Other Positions'. It avoids the possibility to  (simply) add information in this page or to remove or hide the page.

**Comment by kongrattapong** [comment]
@luffah I'm really sorry for late reply.
This module is migration from V11 to V12.
Do we need add a name or something else in migration? 

Thank for your review.

**Review by rousseldenis — CHANGES_REQUESTED** [review]
@kongrattapong You need to preserve commit history. See https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-12.0

**Comment by luffah** [comment]
> Do we need add a name or something else in migration?

No, i imagine. It was just a note for a future improvement. :wink:

**Comment by kongrattapong** [comment]
@rousseldenis Thank you for your review. Fixed it

@luffah I add name attribute as you suggested, please take a look. Thank for your review.

**Comment by rousseldenis** [comment]
> @rousseldenis Thank you for your review. Fixed it
> 
> @luffah I add name attribute as you suggested, please take a look. Thank for your review.

Nearly. I think you have a duplicate commit at the beginning.

**Comment by kongrattapong** [comment]
> Nearly. I think you have a duplicate commit at the beginning.

Thank you, fixed it.

**Comment by rousseldenis** [comment]
And the last one, a flake8 error in travis.

**Review by rousseldenis — APPROVED** [review]
Code review

**Comment by rousseldenis** [comment]
@NL66278 For migration PR, don't use 'major' option of git bot merge.

Module version becomes 12.0.2.0.0 in this case : https://github.com/OCA/partner-contact/blob/12.0/partner_contact_in_several_companies/__manifest__.py#L6

**Comment by NL66278** [comment]
@rousseldenis I can choose between patch , minor and major. In my reasoning a migration is something major. What would you choose?

**Comment by rousseldenis** [comment]
> @rousseldenis I can choose between patch , minor and major. In my reasoning a migration is something major. What would you choose?

In case of migration, just call 'ocabot merge' without any option as the PR is always done (in 99% of cases) with a correct version (e.g.: 12.0.1.0.0)

**Comment by NL66278** [comment]
@rousseldenis OK, thanks. The documentation here: https://odoo-community.org/blog/the-oca-blog-1/post/the-oca-github-bot-can-now-merge-pull-requests-92 led me to think I did need to choose one of the options. Maybe this could be clarified on the OCA website?

**Pull request merged**